macaroon.js

```diff
@@ -338,197 +392,249 @@ const readFieldV2Optional = function(buf, maybeFieldType) {
 */
-function bindForRequest(rootSig, dischargeSig) {
-  if (sjcl.bitArray.equal(rootSig, dischargeSig)) {
-    return rootSig;
+const bindForRequest = function(rootSigBits, dischargeSigBits) {
+  if (sjcl.bitArray.equal(rootSigBits, dischargeSigBits)) {
+    return rootSigBits;
   }
-  return keyedHash2(zeroKey, rootSig, dischargeSig);
-}
+  return keyedHash2(zeroKeyBits, rootSigBits, dischargeSigBits);
+};
 
 const Macaroon = class Macaroon {
   /**
     Create a new Macaroon with the given root key, identifier, location
     and signature.
     @param {Object} The necessary values to generate a macaroon.
       It contains the following fields:
-        identifier: {String}
-        location:   {String}
-        caveats:    {Array}
-        signature:  {bitarray}
+        identifierBytes: {Uint8Array}
+        locationStr:   {string}
+        caveats:    {Array of {locationStr: string, identifierBytes: Uint8Array, vidBytes: Uint8Array}}
+        signatureBytes:  {Uint8Array}
+        version: {int} The version of macaroon to create.
   */
-  constructor({identifier, location, caveats, signature}) {
-    this._location = location;
-    this._identifier = identifier;
-    this._signature = signature;
-    this._caveats = caveats;
+  constructor(params) {
+    if (!params) {
+      // clone uses null parameters.
+      return;
+    }
+    var {version, identifierBytes, locationStr, caveats, signatureBytes} = params;
+    if (version !== 1 && version !== 2) {
+      throw new Error(`Unexpected version ${version}`);
+    }
+    this._version = version;
+    this._locationStr = locationStr;
+    identifierBytes = requireBytes(identifierBytes, 'Identifier');
+    if (version === 1 && !isValidUTF8(identifierBytes)) {
+      throw new Error('Version 1 macaroon identifier must be well-formed UTF-8');
+    }
+    this._identifierBits = identifierBytes && bytesToBits(identifierBytes);
+    this._signatureBits = signatureBytes && bytesToBits(requireBytes(signatureBytes, 'Signature'));
+    this._caveats = caveats ? caveats.map(cav => {
+      const identifierBytes = requireBytes(cav.identifierBytes, 'Caveat identifier');
+      if (version === 1 && !isValidUTF8(identifierBytes)) {
+        throw new Error('Version 1 caveat identifier must be well-formed UTF-8');
+      }
+      return {
+        _locationStr: maybeString(cav.locationStr),
+        _identifierBits: bytesToBits(identifierBytes),
+        _vidBits: cav.vidBytes && bytesToBits(requireBytes(cav.vidBytes, 'Verification ID')),
+      };
+    }) : [];
   }
 
-  get location() {
-    return this._location;
+  /**
+   * Return the caveats associated with the macaroon,
+   * as an array of caveats. A caveat is represented
+   * as an object with an identifier field (Uint8Array)
+   * and (for third party caveats) a location field (string),
+   * and verification id (Uint8Array).
+   * @return {Array} The macaroon's caveats.
+   */
+  get caveats() {
+    if (!this._caveats) {
```

Comment:
Would there ever *not* be a caveat array? Seems you define an empty one in the constructor. .map will just return the empty array if it's empty.

Reply:
good point. done.